### ISSUE TITLE
Is there any ways to create evaluation metrics?

### ISSUE THREAD

**Issue by TrinhTUHH** [tflearn/tflearn#894]
Hi,
I'm building a CNN for grasp detection and I want to use the rectangle metric when evaluating the data. Rectangle metric means both of the following conditions are satisfied:
1) the grasp angle is within 30 degree of the ground truth grasp.
2)  The  Jaccard  index  of  the  predicted  grasp  and  the ground truth is greater than 25 percent.
So my question is how could I create my own metric like this for evaluation?

**Comment by TrinhTUHH** [comment]
So I have tried implementing the metric. 
It looks like this. The grasp_metric is class implemented mainly with numpy and is used for calculating the accuracy of evaluation. 
```
import grasp_metric
def rect_metric(prediction, target, inputs):
	with tf.Session() as sess:
	    pred = prediction.eval(session=sess)
	    tar =  target.eval(session=sess)
        return grasp_error(pred,tar)
network = ...
network = regression(network, metric=rect_metric, optimizer='momentum',
                     loss='mean_square',
                     learning_rate=0.0005)
```
But I got an error like below. 
```
Caused by op u'is_training/read', defined at:
  File "gnet.py", line 11, in <module>
    import tflearn
  File "/usr/local/lib/python2.7/dist-packages/tflearn/__init__.py", line 73, in <module>
    config.init_training_mode()
  File "/usr/local/lib/python2.7/dist-packages/tflearn/config.py", line 128, in init_training_mode
    trainable=False)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/framework/python/ops/arg_scope.py", line 181, in func_with_args
    return func(*args, **current_args)
  File "/usr/local/lib/python2.7/dist-packages/tflearn/variables.py", line 65, in variable
    validate_shape=validate_shape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 1049, in get_variable
    use_resource=use_resource, custom_getter=custom_getter)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 948, in get_variable
    use_resource=use_resource, custom_getter=custom_getter)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 356, in get_variable
    validate_shape=validate_shape, use_resource=use_resource)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 341, in _true_getter
    use_resource=use_resource)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 714, in _get_single_variable
    validate_shape=validate_shape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variables.py", line 197, in __init__
    expected_shape=expected_shape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variables.py", line 316, in _init_from_args
    self._snapshot = array_ops.identity(self._variable, name="read")
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 1338, in identity
    result = _op_def_lib.apply_op("Identity", input=input, name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 768, in apply_op
    op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 2336, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 1228, in __init__
    self._traceback = _extract_stack()

FailedPreconditionError (see above for traceback): Attempting to use uninitialized value is_training
	 [[Node: is_training/read = Identity[T=DT_BOOL, _class=["loc:@is_training"], _device="/job:localhost/replica:0/task:0/cpu:0"](is_training)]]

Reminder: Custom metric function arguments must be defined as: custom_metric(y_pred, y_true, x).
```
There is probably a mistake in the function defining the metric but I don't know how to fix it. Does anyone have ideas?

**Comment by Enet4** [comment]
This doesn't seem to be an issue with tflearn. You just have to implement those operations in TensorFlow, but we do not have enough information about your use case. If you are willing to show us what you have tried so far, including how you implemented `grasp_error` in Python and what exactly is troubling you during this migration, that could turn into a question on [Stack Overflow](https://www.stackoverflow.com).

**Comment by TrinhTUHH** [comment]
So I have uploaded 2 files. The gnet.py is my network which is based on AlexNet and the grasp_metric.py is the one contain graps_error function. I'm quite new to TensorFlow and I did post my question on StackOverflow but no answers so far :(. So please help me take a look.
[GCNN.zip](https://github.com/tflearn/tflearn/files/1284891/GCNN.zip)

**Comment by Enet4** [comment]
Actually, it's best that this is taken over to Stack Overflow completely. It provides more exposure and serves as an incentive to produce a minimal, complete and verifiable example of your problem.

**Comment by TrinhTUHH** [comment]
Ok, thank you. I will do that. By the way, if I implement the metric in TensorFlow instead of using numpy, I still have to run sess.run() or sess.val() every time I want to work with tensor/array, right? Then what makes it different from what I've done so far?

**Comment by TrinhTUHH** [comment]
I repost the question here: https://stackoverflow.com/questions/46105606/how-to-create-a-custom-metric-in-tflearn-or-tensorflow

**Comment by Enet4** [comment]
It is ill-advised to repost the same question. You should have edited the old one into what you now have.

I will now refrain from posting here unless it's deemed necessary.

**Comment by TrinhTUHH** [comment]
I already deleted the old one though!

**Comment by Enet4** [comment]
> I already deleted the old one though!

https://meta.stackoverflow.com/q/265233/1233251
